File: backend/app/api/notifications.py
from typing import Any, Dict, List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query, WebSocket, WebSocketDisconnect

from .auth import get_current_user
from ..services.auth import get_user_from_session
from ..services.notifications import (
    delete_notification,
    get_unread_notification_count,
    get_user_notifications,
    mark_all_notifications_read,
    mark_notification_read,
)
from ..services.websocket import manager

router = APIRouter()


async def get_current_user_ws(websocket: WebSocket) -> Optional[Dict[str, Any]]:
    token = websocket.cookies.get("mkc_session")
    if not token:
        token = websocket.query_params.get("token")
    if not token:
        await websocket.close(code=1008)
        return None
    user = get_user_from_session(token)
    if not user:
        await websocket.close(code=1008)
        return None
    return user


@router.get("", response_model=List[Dict[str, Any]])
async def get_notifications(
    limit: int = Query(default=50, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    current_user: Dict[str, Any] = Depends(get_current_user)
) -> List[Dict[str, Any]]:
    user_id = current_user["id"]
    return get_user_notifications(user_id=user_id, limit=limit, offset=offset)


@router.get("/unread-count", response_model=Dict[str, int])
async def get_unread_count(
    current_user: Dict[str, Any] = Depends(get_current_user)
) -> Dict[str, int]:
    user_id = current_user["id"]
    count = get_unread_notification_count(user_id=user_id)
    return {"unread_count": count}


@router.patch("/read_all", response_model=Dict[str, Any])
@router.patch("/read-all", response_model=Dict[str, Any])
async def mark_all_read(
    current_user: Dict[str, Any] = Depends(get_current_user)
) -> Dict[str, Any]:
    user_id = current_user["id"]
    updated = mark_all_notifications_read(user_id=user_id)
    return {"status": "success", "message": "All notifications marked as read", "count": updated}


@router.patch("/{notification_id}/read", response_model=Dict[str, str])
async def mark_single_read(
    notification_id: int,
    current_user: Dict[str, Any] = Depends(get_current_user)
) -> Dict[str, str]:
    user_id = current_user["id"]
    success = mark_notification_read(user_id=user_id, notification_id=notification_id)
    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")
    return {"status": "success", "message": "Notification marked as read"}


@router.delete("/{notification_id}", response_model=Dict[str, Any])
async def delete_single_notification(
    notification_id: int,
    current_user: Dict[str, Any] = Depends(get_current_user)
) -> Dict[str, Any]:
    user_id = current_user["id"]
    success = delete_notification(user_id=user_id, notification_id=notification_id)
    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")
    return {"status": "success", "message": "Notification deleted", "id": notification_id}


@router.websocket("/ws")
async def notifications_websocket_endpoint(websocket: WebSocket):
    user = await get_current_user_ws(websocket)
    if not user:
        return

    user_id = user["id"]
    await manager.connect(websocket, user_id)
    try:
        while True:
            # Keep socket alive and handle client ping/messages
            data = await websocket.receive_text()
            if data == "ping":
                await websocket.send_json({"type": "pong"})
    except WebSocketDisconnect:
        manager.disconnect(websocket, user_id)

File: netlify/functions/src/controllers/notifications.ts
import { toInt } from '../utils/helpers';
import { Router, Response } from 'express';
import { query } from '../db';
import { AuthenticatedRequest, requireAuth } from '../middleware/auth';

const router = Router();

// ----------------------------------------------------------------------------
// GET /api/notifications
// ----------------------------------------------------------------------------
router.get('/', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const result = await query(
      `SELECT id, user_id, type, title, message, reference_type, reference_id, data, is_read, created_at 
       FROM notifications 
       WHERE user_id = $1 
       ORDER BY created_at DESC 
       LIMIT 50`,
      [req.user!.id]
    );

    return res.json({ notifications: result.rows });
  } catch (err: any) {
    return res.status(500).json({ detail: err?.message || 'Failed to fetch notifications' });
  }
});

// ----------------------------------------------------------------------------
// GET /api/notifications/unread-count
// ----------------------------------------------------------------------------
router.get('/unread-count', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const result = await query('SELECT COUNT(*) FROM notifications WHERE user_id = $1 AND is_read = 0', [
      req.user!.id,
    ]);
    return res.json({ unread_count: parseInt(result.rows[0].count, 10) });
  } catch (err: any) {
    return res.status(500).json({ detail: err?.message || 'Failed to fetch count' });
  }
});

// ----------------------------------------------------------------------------
// PATCH /api/notifications/read-all & /read_all
// ----------------------------------------------------------------------------
const readAllHandler = async (req: AuthenticatedRequest, res: Response) => {
  try {
    await query('UPDATE notifications SET is_read = 1 WHERE user_id = $1', [req.user!.id]);
    return res.json({ message: 'All notifications marked as read' });
  } catch (err: any) {
    return res.status(500).json({ detail: err?.message || 'Failed to mark notifications read' });
  }
};

router.patch('/read-all', requireAuth, readAllHandler);
router.patch('/read_all', requireAuth, readAllHandler);

// ----------------------------------------------------------------------------
// PATCH /api/notifications/:id/read
// ----------------------------------------------------------------------------
router.patch('/:id/read', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const id = toInt(req.params.id);
    await query('UPDATE notifications SET is_read = 1 WHERE id = $1 AND user_id = $2', [id, req.user!.id]);
    return res.json({ message: 'Notification marked as read' });
  } catch (err: any) {
    return res.status(500).json({ detail: err?.message || 'Failed to update notification' });
  }
});

// ----------------------------------------------------------------------------
// DELETE /api/notifications/:id
// ----------------------------------------------------------------------------
router.delete('/:id', requireAuth, async (req: AuthenticatedRequest, res: Response) => {
  try {
    const id = toInt(req.params.id);
    await query('DELETE FROM notifications WHERE id = $1 AND user_id = $2', [id, req.user!.id]);
    return res.json({ message: 'Notification deleted' });
  } catch (err: any) {
    return res.status(500).json({ detail: err?.message || 'Failed to delete notification' });
  }
});

export default router;
